content/blog/2024/2024-02-13-apache-http-client-classic-apis.md

```diff
@@ -0,0 +1,776 @@
+    
+  try (CloseableHttpClient httpClient = HttpClients.createDefault()) {
+    
+    HttpHost httpHost = HttpHost.create("https://reqres.in");
+    
+    URI uri = new URIBuilder("/api/users/").build();
+    HttpOptions httpOptionsRequest = new HttpOptions(uri);
+    
+    // Create a response handler, lambda to implement
+    // HttpClientResponseHandler::handleResponse(ClassicHttpResponse response)
+    HttpClientResponseHandler<Map<String, String>> handler =
+        response ->
+            StreamSupport.stream(
+                    Spliterators.spliteratorUnknownSize(
+                        response.headerIterator(), Spliterator.ORDERED),
+                    false)
+                .collect(Collectors.toMap(Header::getName, Header::getValue));
+    
+    return httpClient.execute(httpHost, httpOptionsRequest, handler);
+  } catch (Exception e) {
+    throw new RequestProcessingException("Failed to execute the request.", e);
+  }
+}
+```
+In this example, it populates the `HttpOptions` request and calls HttpClient `execute()` method. The response from the server is processed by the handler, and the resulting map of headers is returned to the caller.
+
+Now Let's see how to call this functionality.
+
+```java
+@Test
+void executeOptions() {
+  try {
+    // execute
+    Map<String, String> headers = userHttpRequestHelper.executeOptions();
+    assertThat(headers.keySet())
+        .as("Headers do not contain allow header")
+        .containsAnyOf("Allow", "Access-Control-Allow-Methods");
+  } catch (Exception e) {
+    Assertions.fail("Failed to execute HTTP request.", e);
+  }
+}
+```
+The test calls `executeOptions()` to perform the `OPTIONS` request and retrieve the headers from the server. Then it verifies that the keys of the 'headers' map contain at least one of the expected headers ('Allow' or 'Access-Control-Allow-Methods'). 
+
+### Executing HTTP `TRACE` Request to Perform Diagnosis
+
+The HTTP `TRACE` method performs a message loop-back test along the path to the target resource, providing a useful debugging mechanism. However, it is advised not to use this method as it can open the gates to the intruders.
+
+
+{{% danger title="The Vulnerability of TRACE" %}}
+As warned by OWASP in the documentation on [Test HTTP Methods](https://owasp.org/www-project-web-security-testing-guide/latest/4-Web_Application_Security_Testing/02-Configuration_and_Deployment_Management_Testing/06-Test_HTTP_Methods) the `TRACE` method, or `TRACK` in Microsoft's systems, makes the server repeat what it receives in a request. This caused a problem known as `Cross-Site Tracing (XST)` in 2003, allowing access to cookies marked with the `HttpOnly` flag. Browsers and plugins have blocked `TRACE` for years, so this problem is no longer a risk. However, if a server still allows `TRACE`, it might indicate security weaknesses.
+{{% /danger %}}
+
+### Using User Defined Type in Request Processing
```

Comment:
This section is quite long, is htere a way to divide it into subsection?

Reply:
I added sub-sections, and shortened the text also.